File: mobile/lib/models/rental_listing.dart
class RentalListing {
  final int id;
  final int userId;
  final String rentalType; // house, car, commercial, room
  final String title;
  final String? description;
  final String slug;
  final double pricePerUnit;
  final String priceUnit; // hour, day, month, year
  final double securityDeposit;
  final double cleaningFee;
  final String addressLine1;
  final String? addressLine2;
  final String city;
  final String state;
  final String pincode;
  final double? latitude;
  final double? longitude;
  final String status; // draft, active, paused, closed
  final bool instantBooking;
  final List<String>? blockedDates;
  final List<String>? photos;
  final List<String>? rules;
  final int totalBookings;
  final double avgRating;
  final int reviewCount;
  final Map<String, dynamic>? details;
  final Map<String, dynamic>? owner;
  final DateTime createdAt;
  final DateTime updatedAt;

  RentalListing({
    required this.id,
    required this.userId,
    required this.rentalType,
    required this.title,
    this.description,
    required this.slug,
    required this.pricePerUnit,
    required this.priceUnit,
    this.securityDeposit = 0,
    this.cleaningFee = 0,
    required this.addressLine1,
    this.addressLine2,
    required this.city,
    required this.state,
    required this.pincode,
    this.latitude,
    this.longitude,
    this.status = 'draft',
    this.instantBooking = false,
    this.blockedDates,
    this.photos,
    this.rules,
    this.totalBookings = 0,
    this.avgRating = 0,
    this.reviewCount = 0,
    this.details,
    this.owner,
    required this.createdAt,
    required this.updatedAt,
  });

  factory RentalListing.fromJson(Map<String, dynamic> json) {
    return RentalListing(
      id: json['id'] as int,
      userId: json['user_id'] as int,
      rentalType: json['rental_type'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      slug: json['slug'] as String,
      pricePerUnit: (json['price_per_unit'] as num).toDouble(),
      priceUnit: json['price_unit'] as String,
      securityDeposit: (json['security_deposit'] as num?)?.toDouble() ?? 0,
      cleaningFee: (json['cleaning_fee'] as num?)?.toDouble() ?? 0,
      addressLine1: json['address_line1'] as String,
      addressLine2: json['address_line2'] as String?,
      city: json['city'] as String,
      state: json['state'] as String,
      pincode: json['pincode'] as String,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      status: json['status'] as String? ?? 'draft',
      instantBooking: json['instant_booking'] as bool? ?? false,
      blockedDates: json['blocked_dates'] != null
          ? List<String>.from(json['blocked_dates'])
          : null,
      photos: json['photos'] != null
          ? List<String>.from(json['photos'])
          : null,
      rules: json['rules'] != null
          ? List<String>.from(json['rules'])
          : null,
      totalBookings: json['total_bookings'] as int? ?? 0,
      avgRating: (json['avg_rating'] as num?)?.toDouble() ?? 0,
      reviewCount: json['review_count'] as int? ?? 0,
      details: json['details'] as Map<String, dynamic>?,
      owner: json['owner'] as Map<String, dynamic>?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  String get formattedPrice => '₹${pricePerUnit.toStringAsFixed(0)} / $priceUnit';

  String get typeIcon {
    switch (rentalType) {
      case 'house': return '🏠';
      case 'car': return '🚗';
      case 'commercial': return '🏢';
      case 'room': return '🛏️';
      default: return '📦';
    }
  }

  String get typeLabel {
    switch (rentalType) {
      case 'house': return 'House';
      case 'car': return 'Car';
      case 'commercial': return 'Commercial';
      case 'room': return 'Room';
      default: return rentalType;
    }
  }
}

File: app/Models/RentalListing.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Casts\Attribute;
use Illuminate\Database\Eloquent\Relations\HasOne;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class RentalListing extends Model
{
    use SoftDeletes;

    protected $table = 'rental_listings';

    protected $fillable = [
        'user_id', 'rental_type', 'title', 'description', 'slug',
        'price_per_unit', 'price_unit', 'security_deposit', 'cleaning_fee',
        'address_line1', 'address_line2', 'city', 'state', 'pincode',
        'latitude', 'longitude', 'status', 'instant_booking',
        'blocked_dates', 'photos', 'rules',
    ];

    protected $casts = [
        'price_per_unit' => 'decimal:2',
        'security_deposit' => 'decimal:2',
        'cleaning_fee' => 'decimal:2',
        'blocked_dates' => 'array',
        'photos' => 'array',
        'rules' => 'array',
    ];

    // ── Relationships ──────────────────────────────

    public function owner(): BelongsTo
    {
        return $this->belongsTo(User::class, 'user_id');
    }

    public function bookings(): HasMany
    {
        return $this->hasMany(RentalBooking::class, 'rental_listing_id');
    }

    public function reviews(): HasMany
    {
        return $this->hasMany(RentalReview::class, 'rental_listing_id');
    }

    public function availability(): HasMany
    {
        return $this->hasMany(RentalAvailability::class, 'rental_listing_id');
    }

    // ── Type-specific relations ────────────────────

    public function houseDetails(): HasOne
    {
        return $this->hasOne(HouseDetail::class, 'rental_listing_id');
    }

    public function carDetails(): HasOne
    {
        return $this->hasOne(CarDetail::class, 'rental_listing_id');
    }

    public function commercialDetails(): HasOne
    {
        return $this->hasOne(CommercialDetail::class, 'rental_listing_id');
    }

    public function roomDetails(): HasOne
    {
        return $this->hasOne(RoomDetail::class, 'rental_listing_id');
    }

    // ── Accessors ──────────────────────────────────

    protected function details(): Attribute
    {
        return Attribute::get(fn () => match ($this->rental_type) {
            'house' => $this->houseDetails,
            'car' => $this->carDetails,
            'commercial' => $this->commercialDetails,
            'room' => $this->roomDetails,
        });
    }

    protected function formattedPrice(): Attribute
    {
        return Attribute::get(fn () => '₹' . number_format($this->price_per_unit, 0) . ' / ' . $this->price_unit);
    }

    // ── Availability check ─────────────────────────

    public function isAvailable(string $checkIn, string $checkOut): bool
    {
        $range = collect(range(strtotime($checkIn), strtotime($checkOut), 86400))
            ->map(fn ($ts) => date('Y-m-d', $ts));

        // Check JSON blocked_dates column
        $blocked = $this->blocked_dates ?? [];
        if ($range->intersect($blocked)->isNotEmpty()) {
            return false;
        }

        // Check rental_availability table for blocked dates
        $blockedFromTable = $this->availability()
            ->where('status', 'blocked')
            ->whereIn('date', $range->toArray())
            ->exists();

        if ($blockedFromTable) {
            return false;
        }

        $overlap = $this->bookings()
            ->whereIn('status', ['pending', 'confirmed', 'active'])
            ->where('check_in', '<', $checkOut)
            ->where('check_out', '>', $checkIn)
            ->exists();

        return !$overlap;
    }

    // ── Scopes ─────────────────────────────────────

    public function scopeActive($query)
    {
        return $query->where('status', 'active');
    }

    public function scopeOfType($query, string $type)
    {
        return $query->where('rental_type', $type);
    }

    public function scopeInCity($query, string $city)
    {
        return $query->where('city', $city);
    }

    public function scopePriceBetween($query, float $min, float $max)
    {
        return $query->whereBetween('price_per_unit', [$min, $max]);
    }
}
